oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -192,6 +201,61 @@ String createTokenUrlWithScopes() {
     return tokenUrl.toString();
   }
 
+  /**
+   * Gets the universe domain from the GCE metadata server.
+   *
+   * <p>Returns an explicit universe domain if it was provided during credential initialization.
+   *
+   * <p>Returns the {@link Credentials#GOOGLE_DEFAULT_UNIVERSE} if universe domain endpoint is
+   * unavailable or returns an empty string.
+   *
+   * <p>Otherwise, returns universe domain from GCE metadata service.
+   *
+   * <p>Any above value is cached for the credential lifetime.
+   *
+   * @throws IOException if a call to GCE metadata service was unsuccessful. Check if exception
+   *     implements the {@link Retryable} and {@code isRetryable()} will return true if the
+   *     operation may be retried.
+   * @return string representing a universe domain in the format some-domain.xyz
+   */
+  @Override
+  public String getUniverseDomain() throws IOException {
+    if (!isDefaultUniverseDomain()) {
+      return super.getUniverseDomain();
+    }
+
+    synchronized (this) {
+      if (universeDomainFromMetadata != null) {
+        return universeDomainFromMetadata;
+      }
+
+      universeDomainFromMetadata = getUniverseDomainFromMetadata();
+      return universeDomainFromMetadata;
+    }
+  }
+
+  private String getUniverseDomainFromMetadata() throws IOException {
+    HttpResponse response = getMetadataResponse(getUniverseDomainUrl());
+    int statusCode = response.getStatusCode();
+    if (statusCode == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+      return Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    }
+    if (statusCode != HttpStatusCodes.STATUS_CODE_OK) {
+      IOException cause =
+          new IOException(
+              String.format(
+                  "Unexpected Error code %s trying to get universe domain"
+                      + " from Compute Engine metadata for the default service account: %s",
+                  statusCode, response.parseAsString()));
+      throw new GoogleAuthException(true, cause);
+    }
```

Comment:
Just to confirm: Any non-successful status code besides 404 Not Found is always retryable? Are there any possible non-successful status codes that we should not retry?

Reply:
It could be the case, but so far unknown. We know that current behavior to retry everything - so it is safe to continue to do so. We have an item to investigate MDS retryable errors, but it might happen early in 24.

Reply:
Understood. Thanks!